[21.01] Use nested state tracking with sqlalchemy-mutable

mvdbeek opened the pull request:
That should fix AttributeError: 'MutationList' object has no attribute '_key' for good, hopefully.

In the process I also had to eliminate pickling of datasets, which fails with both sqlalchemy-mutable and sqlalchemy-json.
I also removed most uses of OrderedDict ... they could have been allowed as additional JSONType's, but since they don't offer anything in the current codebase (we don't seem to call `reversed` on them) I decided to just remove them.

mvdbeek committed:
Use nested state tracking with sqlalchemy-mutable

instead of the adapted gist recipe we've been using for a long time,
and which fails with:
```
Traceback (most recent call last):
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/workflow/run.py", line 190, in invoke
    incomplete_or_none = self._invoke_step(workflow_invocation_step)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/workflow/run.py", line 266, in _invoke_step
    use_cached_job=self.workflow_invocation.use_cached_job)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/workflow/modules.py", line 1761, in execute
    workflow_resource_parameters=resource_parameters
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/tools/execute.py", line 103, in execute
    execute_single_job(execution_slice, completed_jobs[i])
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/tools/execute.py", line 74, in execute_single_job
    execution_tracker.record_success(execution_slice, job, result)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/tools/execute.py", line 432, in record_success
    self.job_callback(job)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/workflow/modules.py", line 1759, in <lambda>
    job_callback=lambda job: self._handle_post_job_actions(step, job, invocation.replacement_dict),
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/workflow/modules.py", line 1812, in _handle_post_job_actions
    ActionBox.execute(self.trans.app, self.trans.sa_session, pja, job, replacement_dict)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/jobs/actions/post.py", line 517, in execute
    ActionBox.actions[pja.action_type].execute(app, sa_session, pja, job, replacement_dict)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/jobs/actions/post.py", line 103, in execute
    app.datatypes_registry.change_datatype(dataset_instance, action.action_arguments['newtype'])
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/datatypes/registry.py", line 590, in change_datatype
    data.init_meta(copy_from=data)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 2783, in init_meta
    return self.datatype.init_meta(self, copy_from=copy_from)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/datatypes/data.py", line 171, in init_meta
    dataset.metadata = copy_from.metadata
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 2705, in set_metadata
    self._metadata = self.metadata.make_dict_copy(bunch)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/metadata.py", line 161, in make_dict_copy
    rval[key] = self.spec[key].param.make_copy(value, target_context=self, source_context=to_copy)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/metadata.py", line 281, in make_copy
    return copy.deepcopy(value)
  File "/cvmfs/main.galaxyproject.org/deps/_conda/envs/_galaxy_/lib/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/custom_types.py", line 232, in __deepcopy__
    return MutationList(MutationObj.coerce(self._key, copy.deepcopy(self[:])))
AttributeError: 'MutationList' object has no attribute '_key'
```

mvdbeek committed:
Work around bool not subclassable

mvdbeek committed:
Add sqlalchemy-mutable to data package

mvdbeek committed:
Switch to simpler sqlalchemy-json

sqlalchemy-mutable subclasses types, and not all libraries handle this
well (we would need to add CoercedBool isinstance check to the json
encoder), and pyyaml doesn't work with subclassed built-in types at all.

mvdbeek committed:
Drop legacy metadata mode

And always serialize input dataset metadata using the model exports.
This means we don't have to pickle anymore, which breaks with weakrefs
used by sqlalchemy-json's Mutable class.

mvdbeek committed:
Deal with old jobs

mvdbeek committed:
Fix usage of email_alerts column

mvdbeek committed:
Sanitize imports in set_metadata.py

Importing when needed was done to speed up metadata setting, however all
methods for setting metadata need to import galaxy.model, which takes up
the bulk of import time (apart from setting up mappers, so that one I
left where it was). I don't think scattered imports are worth the small
benefit that can be had by doing imports inline.

mvdbeek committed:
Fix loading LDDAs in model store

mvdbeek committed:
Convert more parameters that take basic types to SimpleJSONType

mvdbeek committed:
Pulsar fix

mvdbeek committed:
Fix model store map over collection output

mvdbeek committed:
Revert "Switch to simpler sqlalchemy-json"

This reverts commit fee1fd34cb9fae37c5cff4214f849b5af3c70283.

nsoranzo commented on lib/galaxy/model/custom_types.py:
```diff
@@ -113,149 +113,15 @@ def compare_values(self, x, y):
-                oldvalue._parents.pop(target.obj(), None)
-            return value
+class JSONType(BaseJSONType):
+    pass
```
I'd move this after the definition of `SimpleJSONType` class, so closer to `Mutable.associate_with(JSONType)`.

Also, can you please add a class docstring?

mvdbeek committed:
Add class docstrings

mvdbeek committed:
Fix gxformat2 roundtripping

mvdbeek committed:
Eliminate unncessary use of OrderedDict

Dictionary sort order is stable since Python 3.6, which is the oldest version we
support.

mvdbeek committed:
Drop strip_control_characters_nested workaround

Pretty sure https://github.com/galaxyproject/galaxy/pull/10390/ fixed
the null issue to start with.

jmchilton commented on lib/galaxy/managers/executables.py:
```diff
@@ -1,6 +1,6 @@
 """Utilities for loading tools and workflows from paths for admin user requests."""
 
-from gxformat2.converter import ordered_load
+import yaml
```
I'm certain I had this in here for a reason. Is it just not needed on Python 3?

mvdbeek replied:
Yep, on python 3.6 dicts became ordered, and since 3.7 that behavior is guaranteed

mvdbeek committed:
Fix pulsar test case

mvdbeek commented on lib/galaxy/jobs/runners/pulsar.py:
```diff
@@ -593,6 +593,8 @@ def get_client(self, job_destination_params, job_id, env=None):
             env=env
         )
+        # Turn MutableDict into standard dict for pulsar consumption
+        job_destination_params = dict(job_destination_params.items())
```
So this could be a bugfix ... or not, but probably it is a bugfix. This used to be a a mutable column, so when in the pulsar client we do https://github.com/galaxyproject/pulsar/blob/68a52084468b48a447a4849498d30ac69b805900/pulsar/client/manager.py#L87 we actually change the job_destination_params in the database.

jmchilton replied:
I did know it updated those, Nate and I hit that multiple times last week. I don't know if we're dependent on those database parameters being updated or not (past John didn't leave any clues why he did this xref https://github.com/galaxyproject/pulsar/commit/e4aeb22bca67ab3d20555b81fa73b33a8390d0f9).

I'm okay with merging this and seeing if something breaks. It should be pretty clear if files_endpoint or something needs to be populated initially for updates to work or something odd like that.
